charmcraft/extensions/_utils.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Extension application helpers."""
+
+import copy
+from pathlib import Path
+from typing import Any, Dict, List, Set, Union, cast
+
+from charmcraft.extensions.extension import Extension
+from charmcraft.extensions.registry import get_extension_class
+
+
+def apply_extensions(project_root: Path, yaml_data: Dict[str, Any]) -> Dict[str, Any]:
+    """Apply all extensions.
+
+    :param dict yaml_data: Loaded, unprocessed charmcraft.yaml
+    :returns: Modified charmcraft.yaml data with extensions applied
+    """
+    # Don't modify the dict passed in
+    yaml_data = copy.deepcopy(yaml_data)
+    declared_extensions: List[str] = cast(List[str], yaml_data.get("extensions", []))
+    if not declared_extensions:
+        return yaml_data
+
+    del yaml_data["extensions"]
+
+    # Process extensions in a consistent order
+    for extension_name in sorted(declared_extensions):
+        extension_class = get_extension_class(extension_name)
+        extension = extension_class(project_root=project_root, yaml_data=copy.deepcopy(yaml_data))
```

Comment:
This question may be more appropriate for @sergiusens since he wrote the snapcraft extensions implementation, but why do we do a deep copy here too? Do we expect the extensions to modify the dictionary or are we just taking precautions in case they do even though they're not supposed to?

Based on the extensions in snapcraft I presume it's the latter. How strict do we want to be about this? (e.g. if we're okay with using static analysis to ensure this, could we decorate `Extension.__init__` with `typing.final`?  Or do we want this to be a strict runtime guarantee?)

Reply:
this is probably legacy at this point.

Reply:
I added `self.yaml_data: Final[Dict[str, Any]] = yaml_data`, but I kept deep copy for now.